configuration_go/abstr/kubernetes/thanos/receive/receive.go

```diff
@@ -0,0 +1,514 @@
+package receive
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+
+	cmdopt "github.com/observatorium/observatorium/configuration_go/abstr/kubernetes/cmdoption"
+	"github.com/observatorium/observatorium/configuration_go/k8sutil"
+	"github.com/observatorium/observatorium/configuration_go/schemas/thanos/log"
+	"github.com/observatorium/observatorium/configuration_go/schemas/thanos/option"
+	"github.com/observatorium/observatorium/configuration_go/schemas/thanos/reqlogging"
+	trclient "github.com/observatorium/observatorium/configuration_go/schemas/thanos/tracing/client"
+	monv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus/prometheus/model/relabel"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	dataVolumeName     string = "data"
+	defaultHTTPPort    int    = 10902
+	defaultGRPCPort    int    = 10901
+	defaultReceivePort int    = 19291
+	defaultNamespace   string = "observatorium"
+	defaultImage       string = "quay.io/thanos/thanos"
+)
+
+type GrpcCompressionType string
+type HashRingAlgorithm string
+
+const (
+	GrpcCompressionSnappy    GrpcCompressionType = "snappy"
+	GrpcCompressionNone      GrpcCompressionType = "none"
+	HashRingAlgorithmHashmod HashRingAlgorithm   = "hashmod"
+	HashRingAlgorithmKetama  HashRingAlgorithm   = "ketama"
+)
+
+// Label represents a single label configuration.
+type Label struct {
+	Key   string
+	Value string
+}
+
+func (l Label) String() string {
+	return l.Key + "=" + l.Value
+}
+
+// HashringConfig represents a single hashring configuration.
+type HashringConfig struct {
```

Comment:
Hmm, while I think this is fine, I worry that we will have to keep copying structs like this a lot. I wonder if we should just import/vendor thanos here then?

Reply:
I agree it is not ideal. The issue with the Thanos version is that it does not implement the Stringer interface to marshal to json. This is needed for cmdopt, because it has no idea about the target being json or yaml for a struct.
And we had to redefine many of these type here already because of missing yaml tags or because they include so exported fields for internal use only...

Reply:
or I could add a tag option for cmdopt defining the target serialisation, but I find it even less ideal.

Reply:
Yeah, probably something we need to deal with in the future.